### DISCUSSION TITLE
Android and iOS Gallery Integration

### DISCUSSION THREAD

**Discussion by domhoward14** [Apparence-io/CamerAwesome#337] [Q&A] [answered]
Hey new to this project, but so far it's been really helpful! One thing that took me a while to understand though was after taking video I couldn't see it in the standard iOS and Android photo galleries/viewers which depend on it being saved in specific location on both devices. 

I'm still a little new to flutter so this could be 100% due to my ignorance, but it seems like a little bit of a missed opportunity that the builder for awesomecamera doesn't allow for easy integration with libraries that are made for saving into these necessary locations in order to be viewed on their respective devices easily. 

The way I got it to work is by updating the library code (which i know is already kinda "hacky") for the package in order to control where the file is saved by modifying the `VideoRecordingCameraState` object's `stopRecording` function. I modified it by just adding one line shown below:

`Future<void> stopRecording() async {
    var currentCapture = cameraContext.mediaCaptureController.value;
    if (currentCapture == null) {
      return;
    }
    await CamerawesomePlugin.stopRecordingVideo();
    _mediaCapture = MediaCapture.success(filePath: currentCapture.filePath);
    await CamerawesomePlugin.setCaptureMode(CaptureMode.video);
    cameraContext.changeState(VideoCameraState.from(cameraContext));
    GallerySaver.saveVideo(basenameWithoutExtension(currentCapture.filePath)); // MY MODIFICATION
  }`
  
  and after adding in the update I now had the ability to save the media using a library made to do so across devices and which easily allows for viewing in their respective photo viewers on android and ios.
  
  So i guess my question after all that is did I unnecessarily do this workaround and there's already a better way i can work with the package to achieve the same results, or would it require some updates to make this happen? 
  
  Thanks in advance to any one willing to help :)!

**Comment by apalala-dev** [comment]
Instead of modifying directly CamerAwesome, you can listen to the pictures taken and react to that:
```dart
StreamSubscription<MediaCapture?>? _captureStateSubsciption;
CameraAwesomeBuilder.awesome(
  // Other params
  previewDecoratorBuilder: (state, _, __) {
    _captureStateSubsciption?.cancel();
    _captureStateSubsciption = state.captureState$.listen((event) {
      if (event?.status == MediaCaptureStatus.success) {
        GallerySaver.saveVideo(basenameWithoutExtension(event.filePath));
      }
    });
    return SizedBox.shrink();
  },
);
```
Note that you can setup the listener in any builder that exposes a `CameraState`.

I don't think we'll add this directly in the lib since it does already a lot of things and this feature can be added easily with the help of other packages.
We may expose a kind of controller to be able to setup the listener in a more proper part of the app as discussed in #319.

**Reply by domhoward14** [reply]
oh wow, yeah that is way cleaner then what I did. I wasn't even aware I could subscribe using the builder like that. Awesome! I really appreciate the feedback. I will test it out, and mark this as answer once complete. Thanks again!

**Comment by domhoward14** [comment]
Your subscribe suggestion worked perfectly! I was able to get it to work with the following code snippet.
```
CameraAwesomeBuilder.custom(
        builder: (cameraModeState, previewSize, previewRect) {
          _captureStateSubsciption ??=
              cameraModeState.captureState$.listen((event) {
            if (event?.status == MediaCaptureStatus.success) {
              String filePath = cameraModeState
                      .cameraContext.mediaCaptureController.value?.filePath ??
                  '';
              GallerySaver.saveVideo(filePath)
                  .then((value) => deleteFile(filePath));
            }
          });
 ```

Only potential issue with this is `cameraContext` is a protected member, and I saw a TODO for this to be made private in the future. If this is so, and the file is dynamically created using the SaveConfig, then how can I properly access the filePath field?

**Reply by apalala-dev** [reply, marked as answer]
The `cameraContext` is intended for private behaviours, but you may use it in some special cases. There are discussions to wether we should expose a kind of controller similar to this `cameraContext` or not.

In your case however, you can just use the `event` parameter passed to the `listen` method. It should have the `filePath` 👍

**Reply by domhoward14** [reply]
Ahh, yeah very nice. Well I guess that solves all my problems. Thanks! And I saw the link you posted to the other discussion. Interesting question. I'm still too much of a noob to really understand how much complexity and problems could arise by offering access via a controller object, but on paper that idea sounds good.
